### PR TITLE
Support decoding multiplexed RRD streams

### PR DESCRIPTION
**Pull request by teh-cmc**
TL;DR: the following is now possible:
```
cat docs/snippets/all/archetypes/*_rust.rrd | rerun -
```

This will of course become more interesting as you build more and more complex CLI pipelines with `rerun rrd`

(Also fixed some missing buffered io while I was around.)

### Checklist
* [x] I have read and agree to [Contributor Guide](https://github.com/rerun-io/rerun/blob/main/CONTRIBUTING.md) and the [Code of Conduct](https://github.com/rerun-io/rerun/blob/main/CODE_OF_CONDUCT.md)
* [x] I've included a screenshot or gif (if applicable)
* [x] I have tested the web demo (if applicable):
  * Using examples from latest `main` build: [rerun.io/viewer](https://rerun.io/viewer/pr/7091?manifest_url=https://app.rerun.io/version/main/examples_manifest.json)
  * Using full set of examples from `nightly` build: [rerun.io/viewer](https://rerun.io/viewer/pr/7091?manifest_url=https://app.rerun.io/version/nightly/examples_manifest.json)
* [x] The PR title and labels are set such as to maximize their usefulness for the next release's CHANGELOG
* [x] If applicable, add a new check to the [release checklist](https://github.com/rerun-io/rerun/blob/main/tests/python/release_checklist)!
* [x] If have noted any breaking changes to the log API in `CHANGELOG.md` and the migration guide

- [PR Build Summary](https://build.rerun.io/pr/7091)
- [Recent benchmark results](https://build.rerun.io/graphs/crates.html)
- [Wasm size tracking](https://build.rerun.io/graphs/sizes.html)

To run all checks from `main`, comment on the PR with `@rerun-bot full-check`.

### CODE REVIEW

File: crates/store/re_log_encoding/src/decoder/mod.rs [hunk 1]
```diff
@@ -132,15 +135,39 @@ pub fn read_options(
...
-    read: R,
+    read: Reader<R>,
     uncompressed: Vec<u8>, // scratch space
     compressed: Vec<u8>,   // scratch space
 }
 
 impl<R: std::io::Read> Decoder<R> {
+    /// Instantiates a new decoder.
+    ///
+    /// This does not support multiplexed streams.
```
Comment (emilk) [inline crates/store/re_log_encoding/src/decoder/mod.rs:164]: We should explain what a multiplexed stream even is. It's a stream containing multiple… recordings? or multiple stores (e.g. a recording and a blueprint)?

File: crates/store/re_log_encoding/src/decoder/mod.rs [hunk 2]
```diff
@@ -132,15 +135,39 @@ pub fn read_options(
...
+            Self::Buffered(read) => read.read(buf),
+        }
+    }
+}
+
 pub struct Decoder<R: std::io::Read> {
     version: CrateVersion,
     compression: Compression,
-    read: R,
+    read: Reader<R>,
```
Comment (emilk) [inline crates/store/re_log_encoding/src/decoder/mod.rs:156]: Do we ever want to support unbuffered reads? Aren't those just strictly slower in almost all cases?

File: crates/store/re_log_encoding/src/decoder/mod.rs (hunk 1 above)
Comment (emilk) [reply crates/store/re_log_encoding/src/decoder/mod.rs:164]: "Multiplexing" sounds to me that the streams are interleaved. I think "concatenated" is the words we're looking for here.

File: crates/store/re_log_encoding/src/decoder/mod.rs [hunk 3]
```diff
@@ -153,7 +180,43 @@ impl<R: std::io::Read> Decoder<R> {
...
+    /// This will gracefully handle multiplexed RRD streams, at the cost of extra performance
+    /// overhead, by looking ahead for potential `FileHeader`s in the stream.
+    ///
+    /// The [`CrateVersion`] of the final, demultiplexed stream will correspond to the most recent
+    /// version among all the versions found in the stream.
+    ///
+    /// This is particularly useful when working with stdio streams.
+    ///
+    /// If you're not familiar with multiplexed RRD streams, then you probably want to use
+    /// [`Decoder::new`] instead.
```
Comment (emilk) [inline crates/store/re_log_encoding/src/decoder/mod.rs:200]: Why do we need both constructors though? They look identical, except one requires a buffered reader.

Isn't it just better if we have one constructor that always handled concatenated streams?

File: crates/store/re_log_encoding/src/decoder/mod.rs (hunk 2 above)
Comment (teh-cmc) [reply crates/store/re_log_encoding/src/decoder/mod.rs:156]: Not all cases -- if you're reading from an array of bytes, like we do in a bunch of places, adding extra buffering is pure waste of time and space.

File: crates/store/re_log_encoding/src/decoder/mod.rs (hunk 1 above)
Comment (teh-cmc) [reply crates/store/re_log_encoding/src/decoder/mod.rs:164]: I don't mind "concatenated".

File: crates/store/re_log_encoding/src/decoder/mod.rs (hunk 3 above)
Comment (teh-cmc) [reply crates/store/re_log_encoding/src/decoder/mod.rs:200]: Three reasons:
1. As mentioned above, buffering is a net loss is some real world cases that we depend on today.
2. There's non negligible overhead to constantly check for unexpected `FileHeader`s, which there's really no reason to be paying for in most cases.
3. This is a very specific constructor that relies specifically on `std::io::BufReader`, as opposed to any type that implements `std::io::BufRead`.
